update/sync/syncAccountsDBs.go

```diff
@@ -0,0 +1,210 @@
+			if err != nil {
+				mutErr.Lock()
+				errFound = err
+				mutErr.Unlock()
+			}
+			wg.Done()
+		}(shData)
+	}
+
+	err := WaitFor(chDone, waitTime)
+	if err != nil {
+		return err
+	}
+
+	if errFound != nil {
+		return errFound
+	}
+
+	st.mutSynced.Lock()
+	st.synced = true
+	st.mutSynced.Unlock()
+
+	return nil
+}
+
+func (st *syncAccountsDBs) syncMeta(meta *block.MetaBlock) error {
+	err := st.syncAccountsOfType(state.UserAccount, state.UserAccountsState, core.MetachainShardId, meta.RootHash)
+	if err != nil {
+		return err
+	}
+
+	err = st.syncAccountsOfType(state.ValidatorAccount, state.PeerAccountsState, core.MetachainShardId, meta.ValidatorStatsRootHash)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (st *syncAccountsDBs) syncShard(shardData block.EpochStartShardData) error {
+	err := st.syncAccountsOfType(state.UserAccount, state.UserAccountsState, shardData.ShardId, shardData.RootHash)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (st *syncAccountsDBs) syncAccountsOfType(accountType state.Type, trieID state.AccountsDbIdentifier, shardId uint32, rootHash []byte) error {
+	accAdapterIdentifier := genesis.CreateTrieIdentifier(shardId, accountType)
+
+	success := st.tryRecreateTrie(shardId, accAdapterIdentifier, trieID, rootHash)
+	if success {
+		return nil
+	}
+
+	accountsDBSyncer, err := st.accountsBDsSyncers.Get(accAdapterIdentifier)
+	if err != nil {
+		return err
+	}
+
+	err = accountsDBSyncer.SyncAccounts(rootHash)
+	if err != nil {
+		// critical error - should not happen - maybe recreate trie syncer here
+		return err
+	}
+
+	st.setTries(shardId, accAdapterIdentifier, rootHash, accountsDBSyncer.GetSyncedTries())
+
+	return nil
+}
+
+func (st *syncAccountsDBs) setTries(shId uint32, initialID string, rootHash []byte, tries map[string]data.Trie) {
+	for hash, currentTrie := range tries {
+		if bytes.Equal(rootHash, []byte(hash)) {
+			st.tries.setTrie(initialID, currentTrie)
+			continue
+		}
+
+		dataTrieIdentifier := genesis.CreateTrieIdentifier(shId, state.DataTrie)
+		identifier := genesis.AddRootHashToIdentifier(dataTrieIdentifier, hash)
+		st.tries.setTrie(identifier, currentTrie)
+	}
+}
+
+func (st *syncAccountsDBs) tryRecreateTrie(shardId uint32, id string, trieID state.AccountsDbIdentifier, rootHash []byte) bool {
+	savedTrie, ok := st.tries.getTrie(id)
+	if ok {
+		currHash, err := savedTrie.Root()
+		if err == nil && bytes.Equal(currHash, rootHash) {
+			return true
+		}
+	}
+
+	activeTrie := st.activeAccountsDBs[trieID]
+	if check.IfNil(activeTrie) {
+		return false
+	}
+
+	tries, err := activeTrie.RecreateAllTries(rootHash)
+	if err != nil {
+		return false
+	}
+
+	for _, recreatedTrie := range tries {
+		err = recreatedTrie.Commit()
+		if err != nil {
+			return false
+		}
+	}
+
+	st.setTries(shardId, id, rootHash, tries)
+
+	return true
+}
+
+// GetTries returns the synced tries
+func (st *syncAccountsDBs) GetTries() (map[string]data.Trie, error) {
+	if !st.synced {
```

Comment:
reading the synced state needs to be protected

Reply:
done.